### PR TITLE
Root Motion Extraction Rule (preview)

### PR DESCRIPTION
**Pull request by amzn-rhhong**
Add a preview version of the root motion extraction rule.
image
image

Video attached
https://user-images.githubusercontent.com/69218254/175829755-4ca04520-9c15-44da-b894-091858b4d745.mp4

Root motion extraction is a way to automatically generate root motion based on sample joint's animation (usually the hip joint). 

This change only focus on the transition, so it won't work for anything that contain rotation (like turning).

### CODE REVIEW

File: Gems/EMotionFX/Code/EMotionFX/Pipeline/RCExt/Motion/MotionDataBuilder.cpp [hunk 1]
```diff
@@ -258,6 +265,16 @@ namespace EMotionFX
...
                 }
 
+                const SceneContainers::SceneGraph::NodeIndex boneNodeIndex = graph.Find(it->first.GetPath());
+                const char* nodeName = it->first.GetName();
+                const char* nodePath = it->first.GetPath();
+
+                // Add root joint to motion data when root motion extraction rule exsits.
+                if (rootMotionExtractionRule && boneNodeIndex == rootBoneNodeIndex)
+                {
+                    rootJointDataIndex = motionData->AddJoint(it->first.GetName(), Transform::CreateIdentity(), Transform::CreateIdentity());
```
Comment (pollend) [inline Gems/EMotionFX/Code/EMotionFX/Pipeline/RCExt/Motion/MotionDataBuilder.cpp:275]: you declared a variable for this`it->first.GetName()` could you replace it with nodeName?

File: Gems/EMotionFX/Code/EMotionFX/Pipeline/SceneAPIExt/Rules/RootMotionExtractionRule.cpp [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+                ->DataElement(AZ::Edit::UIHandlers::Default, &RootMotionExtractionData::m_transitionZeroXAxis, "Force X Axis Transition to 0", "Force X Axis movement to be zero.")
+                ->DataElement(AZ::Edit::UIHandlers::Default, &RootMotionExtractionData::m_transitionZeroYAxis, "Force Y Axis Transition to 0", "Force Y Axis movement to be zero.");
+        }
+    }
+
+    RootMotionExtractionData::RootMotionExtractionData()
+    {
+    }
+
+    ///////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
```
Comment (pollend) [inline Gems/EMotionFX/Code/EMotionFX/Pipeline/SceneAPIExt/Rules/RootMotionExtractionRule.cpp:51]: not sure what you are trying to separate here?

File: Gems/EMotionFX/Code/EMotionFX/Pipeline/RCExt/Motion/MotionDataBuilder.cpp (hunk 1 above)
Comment (amzn-rhhong) [reply Gems/EMotionFX/Code/EMotionFX/Pipeline/RCExt/Motion/MotionDataBuilder.cpp:275]: sure good catch

File: Gems/EMotionFX/Code/EMotionFX/Pipeline/SceneAPIExt/Rules/RootMotionExtractionRule.cpp (hunk 2 above)
Comment (amzn-rhhong) [reply Gems/EMotionFX/Code/EMotionFX/Pipeline/SceneAPIExt/Rules/RootMotionExtractionRule.cpp:51]: between the implementation of RootMotionExtractionData and RootMotionExtractionRule.

File: Gems/EMotionFX/Code/EMotionFX/Pipeline/SceneAPIExt/Rules/RootMotionExtractionRule.cpp [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+        AZ::EditContext* editContext = serializeContext->GetEditContext();
+        if (editContext)
+        {
+            editContext->Class<RootMotionExtractionData>("Root motion extraction data", "Root motion extraction data.")
+                ->ClassElement(AZ::Edit::ClassElements::EditorData, "")
+                    ->Attribute(AZ::Edit::Attributes::AutoExpand, true)
+                ->DataElement("NodeListSelection", &RootMotionExtractionData::m_sampleJoint, "Sample joint", "Sample joint to extract motion data from. Usually the hip joint.")
+                    ->Attribute("ClassTypeIdFilter", AZ::SceneAPI::DataTypes::IBoneData::TYPEINFO_Uuid())
+                ->ClassElement(AZ::Edit::ClassElements::Group, "Transition Extraction")
+                ->DataElement(AZ::Edit::UIHandlers::Default, &RootMotionExtractionData::m_transitionZeroXAxis, "Force X Axis Transition to 0", "Force X Axis movement to be zero.")
```
Comment (amzn-jillich) [inline Gems/EMotionFX/Code/EMotionFX/Pipeline/SceneAPIExt/Rules/RootMotionExtractionRule.cpp:42]: What about "Ignore X-Axis" or "Ignore X-Axis Movement" here instead? Might be more easy to understand for users?

File: Gems/EMotionFX/Code/EMotionFX/Source/MotionData/NonUniformMotionData.cpp [hunk 4]
```diff
@@ -1204,6 +1204,51 @@ namespace EMotionFX
...
+                m_jointData[rootJointDataIndex].m_positionTrack.m_values[i].SetZ(0);
+
+                // Compensation in samples.
+                const float x = transitionZeroXAxis ? m_jointData[sampleJointDataIndex].m_positionTrack.m_values[i].GetX() : 0;
+                const float y = transitionZeroYAxis ? m_jointData[sampleJointDataIndex].m_positionTrack.m_values[i].GetY() : 0;
+                const float z = m_jointData[sampleJointDataIndex].m_positionTrack.m_values[i].GetZ();
+
+                m_jointData[sampleJointDataIndex].m_positionTrack.m_values[i].Set(x, y, z);
+            }
+        }
```
Comment (amzn-jillich) [inline Gems/EMotionFX/Code/EMotionFX/Source/MotionData/NonUniformMotionData.cpp:1239]: Rotation isn't taken into account yet, is this planned for a later iteration?

**Review by amzn-jillich — APPROVED** [review]

File: Gems/EMotionFX/Code/EMotionFX/Pipeline/RCExt/Actor/ActorBuilder.cpp [hunk 5]
```diff
@@ -336,28 +336,19 @@ namespace EMotionFX
...
+            for (auto it = graphDownwardsRootBoneView.begin(); it != graphDownwardsRootBoneView.end(); ++it)
             {
                 const SceneContainers::SceneGraph::NodeIndex& nodeIndex = graph.ConvertToNodeIndex(it.GetHierarchyIterator());
 
                 // The end point in ly scene graph should not be added to the emfx actor.
                 // Note: For example, the end point could be a transform node. We will process that later on its parent node.
                 if (graph.IsNodeEndPoint(nodeIndex))
                 {
-                    continue;
+                    // Skip the end point node except if it's a root node.
```
Comment (hultonha) [inline Gems/EMotionFX/Code/EMotionFX/Pipeline/RCExt/Actor/ActorBuilder.cpp:347]: +1 on the guide comment 🙂

File: Gems/EMotionFX/Code/EMotionFX/Pipeline/SceneAPIExt/Rules/RootMotionExtractionRule.h [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+
+#include <AzCore/Memory/SystemAllocator.h>
+#include <AzCore/std/smart_ptr/unique_ptr.h>
+#include <SceneAPI/SceneCore/DataTypes/Rules/IRule.h>
+#include <SceneAPI/SceneData/SceneDataConfiguration.h>
+#include <SceneAPIExt/Rules/ExternalToolRule.h>
+
+namespace EMotionFX::Pipeline::Rule
+{
+    struct RootMotionExtractionData
```
Comment (hultonha) [inline Gems/EMotionFX/Code/EMotionFX/Pipeline/SceneAPIExt/Rules/RootMotionExtractionRule.h:19]: suggestion: An API comment may well be useful for this public type

File: Gems/EMotionFX/Code/EMotionFX/Pipeline/SceneAPIExt/Rules/RootMotionExtractionRule.h [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+
+        static void Reflect(AZ::ReflectContext* context);
+
+    private:
+        bool m_transitionZeroXAxis = false;
+        bool m_transitionZeroYAxis = false;
+        AZStd::string m_sampleJoint = "Hip";
+    };
+
+    class RootMotionExtractionRule final
```
Comment (hultonha) [inline Gems/EMotionFX/Code/EMotionFX/Pipeline/SceneAPIExt/Rules/RootMotionExtractionRule.h:39]: nit: API comment here would likely be useful

**Review by hultonha — APPROVED** [review]
Overall looks like a fantastic first step, awesome work @amzn-rhhong! 😄 I just have some fairly minor suggestions/comments. The other thing I wanted to ask about are tests 😉 Not strictly necessary for this PR, but we should think now about how to test and verify this new logic/behavior and ensure it's doing what we expect.

**Comment by amzn-rhhong** [comment]
@hultonha Good tips! Some of the suggestion you mentioned already got worked on in the next PR and a few class already been changed. I will directly address the reviews in the coming PR.

**Pull request merged**